server/services/redis.service.js

```diff
@@ -33,6 +38,39 @@ module.exports = class RedisService {
     const keyWithSessionId = `${request.state[DEFRA_IVORY_SESSION_KEY]}.${key}`
     client.del(keyWithSessionId)
   }
+
+  static async deleteSessionData (request) {
+    const client = request.redis.client
+    const sessionKey = request.state[DEFRA_IVORY_SESSION_KEY]
+    const totalPossibleKeys =
+      Object.keys(RedisKeys).length +
+      UploadPhoto.MAX_PHOTOS +
+      UploadDocument.MAX_DOCUMENTS
+
+    const keys = []
+
+    let scanResult = await client.scan('0', 'MATCH', `${sessionKey}.*`)
+    let cursor = scanResult[0]
+
+    keys.push(...scanResult[1])
+
+    while (cursor !== '0') {
+      scanResult = await client.scan(cursor, 'MATCH', `${sessionKey}.*`)
+      cursor = scanResult[0]
+      keys.push(...scanResult[1])
+    }
+
+    if (keys.length > totalPossibleKeys) {
+      // Mitigates against a malicious attack using this wildcard search to remove all Redis keys
```

Comment:
Need to confirm if this is actually needed... seeing as Redis keys start with the session ID

e.g. "b5e1f4f9-31f3-4bfa-8389-311a72c1ce20.submission-reference"

And the session ID is encrypted in the session cookie, how is it possible for more keys to be deleted than contained in the current user session?

Reply:
The wildcard scan/delete on the Redis cache made me a bit nervous. Not that it should be possible, but if the key in the cookie that is passed to the function was somehow altered to a *. There is the potential for the whole Redis cache to be cleared out. Put it in there just in case.